src/renderer/dx/DxFontRenderData.h

```diff
@@ -74,9 +88,29 @@ namespace Microsoft::Console::Render
 
         [[nodiscard]] static HRESULT STDMETHODCALLTYPE s_CalculateBoxEffect(IDWriteTextFormat* format, size_t widthPixels, IDWriteFontFace1* face, float fontScale, IBoxDrawingEffect** effect) noexcept;
 
+        bool DidUserSetFeatures() const noexcept;
+        void SetFeatures(const std::unordered_map<std::wstring_view, uint32_t>& features);
+        void SetAxes(const std::unordered_map<std::wstring_view, int32_t>& axes);
+
+        float FontStretchToWidthAxisValue(DWRITE_FONT_STRETCH fontStretch) noexcept;
+        float FontStyleToSlantFixedAxisValue(DWRITE_FONT_STYLE fontStyle) noexcept;
+        float DIPsToPoints(const float fontSize) noexcept;
+        std::vector<DWRITE_FONT_AXIS_VALUE> GetAxisVector(const DWRITE_FONT_WEIGHT fontWeight,
+                                                          const DWRITE_FONT_STRETCH fontStretch,
+                                                          const DWRITE_FONT_STYLE fontStyle,
+                                                          const float fontSize,
+                                                          IDWriteTextFormat3* format);
+
     private:
         using FontAttributeMapKey = uint32_t;
 
+        bool _didUserSetFeatures{ false };
+        // The font features to apply to the text
+        std::vector<DWRITE_FONT_FEATURE> _featureVector;
+
+        // The font axes to apply to the text
+        std::vector<DWRITE_FONT_AXIS_VALUE> _axesVector;

```

Comment:
@lhecker, do these allocate by default?

We have, in the past, used a `basic_string<T>` when we wanted to take advantage of small-string optimization. Is there value in doing that here? I know it is a perversion of the string interface, but it would be alloc-free and just grow the instance body a little bit.

Reply:
(in the zero feature case.)

Reply:
Sorry I didn't see your comment: `std::vector` doesn't allocate in its default constructor.
Unless the user sets the associated config setting, I wouldn't even push anything into those vectors, which makes it "cost-free".